### ISSUE TITLE
Consistent and simple ./out directory structure

### ISSUE THREAD

**Issue by ReinhardKeil** [Open-CMSIS-Pack/devtools#821]
**Describe The Problem To Be Solved**

The ./out directory is currently complex to understand. The output of a build process is typically the input for post-processing the images which could be Flash-Programming, Signature-Tool, Cloud-based OTA service, etc.  A default setting for the ./out directory should be therefore simple and consistent across the tools.

Proposed structure for multi-core projects (exemplified with [csolution-examples/DualCore](https://github.com/Open-CMSIS-Pack/dual_core))

```txt
./Hello.csoution.yml                                     // location of csolution.yml is the base directory for ./out
./out/FRDM-K32L3A6/HelloWorld_cm0plus.axf
./out/FRDM-K32L3A6/HelloWorld_cm0plus.axf.map
./out/FRDM-K32L3A6/HelloWorld_cm0plus.bin
./out/FRDM-K32L3A6/HelloWorld_cm0plus.hex
./out/FRDM-K32L3A6/HelloWorld_cm4.axf
./out/FRDM-K32L3A6/HelloWorld_cm4.axf.map
./out/FRDM-K32L3A6/HelloWorld_cm4.bin
./out/FRDM-K32L3A6/HelloWorld_cm4.hex
```

In a nutshell, all output artifacts of a target are in the same directory. Each *.cproject.yml file has a project name which is used as output name. 

> NOTE: The build-type is not part of the directory naming, as it would otherwise require to adjust post-processing steps for each different build-type. Build-types are used to create test, debug, and release builds depending on the need of a project. In many cases there might be different build-types used during development, i.e. the tools should make it easy to mix a HelloWorld_cm4 release build with a HelloWorld_cm0plus debug or test build.

> NOTE: This proposal requires to review the [output:](https://github.com/Open-CMSIS-Pack/devtools/blob/main/tools/projmgr/docs/Manual/YML-Input-Format.md#output) node.

**Comment by mcgordonite** [comment]
One thing to consider here is that post-processing tools can (should?) use the `output` node in the generated cbuild.yml files to locate the output binaries. See the discussion here: https://github.com/Open-CMSIS-Pack/devtools/issues/778. This is how the [Csolution extension](https://github.com/Arm-Software/vscode-cmsis-csolution/) locates binaries.

That means the tools don't need to understand how the `output` and `output-dirs` nodes are processed by the build tools, and tools don't need to care if the out directory has a complex structure.

However, I think a simpler output directory structure could help the user to understand what is going on, if we can still support all of the use cases.

**Comment by ReinhardKeil** [comment]
Matt is there a description of the default behavior of the Csolution extension.  Does this consider multi-project solutions?

**Comment by slhultgren** [comment]
I'm worried about the number of potential file conflicts behind `.o` and `CMakeLists.txt` etc if all output files are in the same folder.

On the two points:
`out directory is complex` - I don't really agree. Separating build output per project context really makes it less complex IMO since I can easily connect the dots of where something ends up.

`Post-processing tools should find the elf files` - This is a problem indeed, but I agree with @mcgordonite , it could be solved by the `output` property recorded in the cbuild.yml. If post-processing tools are run as a `post-build-step` we could consider modifying or adding a new access sequence e.g. `$CurrentOutputFile(elf)$` that would expand to the file path for the elf file (in the current context) and could be passed as a parameter to any post build script (or use the existing `$Output($BuildType$$TargetType$)$`).  

New point:
`Finding import library for trustzone` - On this, instead of guessing the import file name, I suggest again that a access sequence should be used here

**Comment by ReinhardKeil** [comment]
@slhultgren there is no change to the `.tmp` directory that holds all the artifacts that you describe above (`.o`, `CMakeLists.txt`, etc.).  So the build process itself has separate directories for each build and target type.

The `output-dirs:` control is today already flexible enough to configure the `out` directory the way described. All that is needed is:

```yml
  output-dirs:
    outdir: ./out/$TargetType$/
```
So you can experiment with it.

**Comment by ReinhardKeil** [comment]
Examples in this https://github.com/Open-CMSIS-Pack/csolution-examples/pull/10/files are not configured this way.

**Comment by ReinhardKeil** [comment]
Replaced by `context-set` discussion
